src/Collateral.sol

```diff
@@ -0,0 +1,202 @@
+ *         from `RAVs`.
+ */
+contract Collateral {
+    struct CollateralAccount {
+        uint256 balance; // Total collateral balance for a sender-receiver pair
+        uint256 amountThawing; // Amount of collateral currently being thawed
+        uint256 thawEnd; // Block number at which thawing period ends
+    }
+
+    // Stores how much collateral each sender has deposited for each receiver, as well as thawing information
+    mapping(address => mapping(address => CollateralAccount)) private collateralAccounts;
+    // Map of authorized signers to which sender they are authorized to sign for
+    // key: signer address, value: sender address
+    mapping(address => address) private authorizedSigners;
+
+    // The ERC20 token used for collateral
+    IERC20 public immutable collateralToken;
+
+    // The contract used for verifying receipt aggregate vouchers
+    TAPVerifier public immutable tapVerifier;
+
+    // The contract used for tracking used allocation IDs
+    AllocationIDTracker public immutable allocationIDTracker;
+
+    // The duration in which collateral funds are thawing before they can be withdrawn
+    uint256 public immutable thawingPeriod;
+
+    /**
+     * @dev Emitted when collateral is deposited for a receiver.
+     */
+    event Deposit(address indexed sender, address indexed receiver, uint256 amount);
+
+    /**
+     * @dev Emitted when collateral is redeemed by a receiver.
+     */
+    event Redeem(address indexed receiver, uint256 amount);
+
+    /**
+     * @dev Emitted when a thaw request is made for collateral.
+     */
+    event ThawRequest(address indexed sender, address indexed receiver, uint256 amount, uint256 thawEnd);
+
+    /**
+     * @dev Emitted when thawed collateral is withdrawn by the sender.
+     */
+    event Withdraw(address indexed sender, address indexed receiver, uint256 amount);
+
+    /**
+     * @dev Emitted when a signer is authorized to sign RAVs for a sender.
+     */
+    event AuthorizeSigner(address indexed signer, address indexed sender);
+
+    constructor(address collateralToken_, address tapVerifier_, address allocationIDTracker_, uint256 thawingPeriod_) {
+        collateralToken = IERC20(collateralToken_);
+        tapVerifier = TAPVerifier(tapVerifier_);
+        allocationIDTracker = AllocationIDTracker(allocationIDTracker_);
+        thawingPeriod = thawingPeriod_;
+    }
+
+    /**
+     * @dev Deposits collateral for a receiver.
+     * @param receiver Address of the receiver.
+     * @param amount Amount of collateral to deposit.
+     * @notice The collateral must be approved for transfer by the sender.
+     * @notice REVERT: this function will revert if the collateral transfer fails.
+     */
+    function depositCollateral(address receiver, uint256 amount) external {
+        collateralAccounts[msg.sender][receiver].balance += amount;
+        emit Deposit(msg.sender, receiver, amount);
```

Comment:
In general events should be emitted after all processing is done. This ensures a correct order of events in case there are interactions in the method call that also emit events. 

For example in this case the `Deposit` would be logged before the `Transfer` on the next line, which is not quite true since the deposit is only completed after the transfer is successful.

There are several instances of events being emitted before contract interactions in this contract.

Reply:
I added this change but realized after why it was done in that order. I am using slither for linting and it gives the following detector: https://github.com/crytic/slither/wiki/Detector-Documentation#reentrancy-vulnerabilities-3

Reply:
That's interesting, first time I see that recommendation. I understand where it's coming from, though I'd argue that you shouldn't be designing for well ordered re-entrancy events but rather what makes sense. 

To me a `Deposit` event (to cite an example) signals that a deposit has been finalized and that only happens _after_ the token transfer has been completed. So the event order should be `Transfer` then `Deposit`. But then again, this is probably something developers could start a war over 😆

Reply:
Yeah I agree with your thought process. If the intent of the event was to show a deposit is being started then it would make sense to have it before, but this event is to show when it is being completed. I think ordering of events is more important than handling a re-entrancy that we are not expecting in the first place.